JPA-Annotation @AttributeOverride is not evaluated properly

### Describe the bug

I have an application with Quarkus using JPA, which has the following problem since I migrated from Quarkus 1.x to 2.x.

There exists a MappedSuperclass named BaseClass, with the following field, which is mapped by default to the database-column 'name':
`
private String name;
`
Now I have another class named DerivedClass, where I want to map the field to another db-column 'custom_name'.
`@Entity
@AttributeOverride(name = "name", column = @Column(name = "custom_name"))  
public class DerivedClass extends BaseClass`

But apperently the AttributeOverride-Annotation is ignored at runtime, as the application tries to read the (not existing) db-column 'name' - and not the column 'custom_name', as required by the annotation.

I managed to fix the problem via the following entry in the orm.xml-file, but I wonder why the AttributeOverride-Annotation is broken since Quarkus 2.x

`
    <entity class="DerivedClass">  
        <attribute-override name="name">  
            <column name="custom_name"/>  
        </attribute-override>  
    </entity>
`

### Expected behavior

Expected behaviour is that the database column custom_name is read.

### Actual behavior

2021-09-02 14:51:52,332 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (executor-thread-0) SQL Error: 0, SQLState: 42703
2021-09-02 14:51:52,333 ERROR [org.hib.eng.jdb.spi.SqlExceptionHelper] (executor-thread-0) ERROR: column derivedclass2_.name does not exist

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "11.0.10" 2021-01-19 LTS

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.2.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.1

### Additional information

_No response_

@yrodiere 
I have modified the Reproducer so that the test now fails:
```
[ERROR]   AttributeOverrideTest.attributeOverrideTakenIntoAccount:48
Expecting HashSet:
  ["name"]
to contain:
  ["custom_name"]
but could not find the following element(s):
  ["custom_name"]
```

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -90,7 +90,7 @@
         <commons-lang3.version>3.12.0</commons-lang3.version>
         <commons-codec.version>1.15</commons-codec.version>
         <classmate.version>1.5.1</classmate.version>
-        <hibernate-orm.version>5.6.0.Beta1</hibernate-orm.version>
+        <hibernate-orm.version>5.6.0.Beta2</hibernate-orm.version>
         <hibernate-reactive.version>1.0.0.CR9</hibernate-reactive.version>
         <hibernate-validator.version>6.2.0.Final</hibernate-validator.version>
         <hibernate-search.version>6.0.6.Final</hibernate-search.version>

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmAnnotations.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmAnnotations.java
@@ -138,6 +138,7 @@ public final class HibernateOrmAnnotations {
             DotName.createSimple("org.hibernate.annotations.ColumnTransformer"),
             DotName.createSimple("org.hibernate.annotations.ColumnTransformers"),
             DotName.createSimple("org.hibernate.annotations.Columns"),
+            DotName.createSimple("org.hibernate.annotations.Comment"),
             DotName.createSimple("org.hibernate.annotations.CreationTimestamp"),
             DotName.createSimple("org.hibernate.annotations.DiscriminatorFormula"),
             DotName.createSimple("org.hibernate.annotations.DiscriminatorOptions"),

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/enhancer/HibernateEntityEnhancerFinalFieldTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/enhancer/HibernateEntityEnhancerFinalFieldTest.java
@@ -1,0 +1,332 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package io.quarkus.hibernate.orm.enhancer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.UUID;
+
+import javax.inject.Inject;
+import javax.persistence.Embeddable;
+import javax.persistence.Embedded;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import javax.persistence.EntityManager;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.transaction.NotSupportedException;
+import javax.transaction.SystemException;
+import javax.transaction.UserTransaction;
+
+import org.hibernate.annotations.Immutable;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Checks that final field write access is not replaced with a call to an internal Hibernate ORM method
+ * (since writes to final fields can only occur from constructors).
+ * <p>
+ * See https://github.com/quarkusio/quarkus/issues/20186
+ */
+public class HibernateEntityEnhancerFinalFieldTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(
+                            EntityWithFinalField.class,
+                            EntityWithEmbeddedIdWithFinalField.class, EntityWithEmbeddedIdWithFinalField.EmbeddableId.class,
+                            EntityWithEmbeddedNonIdWithFinalField.class,
+                            EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId.class))
+            .withConfigurationResource("application.properties");
+
+    @Inject
+    EntityManager em;
+
+    @Inject
+    UserTransaction transaction;
+
+    @Test
+    public void entityWithFinalField_constructor() {
+        EntityWithFinalField entity = new EntityWithFinalField("foo");
+        assertThat(entity.immutableProperty).isEqualTo("foo");
+    }
+
+    // Just test that the embedded non-ID works correctly over a persist/retrieve cycle
+    @Test
+    public void entityWithFinalField_smokeTest() {
+        EntityWithFinalField persistedEntity = new EntityWithFinalField("foo");
+        persistedEntity.setName("Some name");
+        inTransaction(() -> {
+            em.persist(persistedEntity);
+        });
+
+        inTransaction(() -> {
+            EntityWithFinalField entity = em.find(EntityWithFinalField.class, persistedEntity.getId());
+            assertThat(entity).extracting(EntityWithFinalField::getImmutableProperty)
+                    .isEqualTo(persistedEntity.getImmutableProperty());
+        });
+    }
+
+    // Just test that the embedded ID works correctly over a persist/retrieve cycle
+    @Test
+    public void embeddableIdWithFinalField_smokeTest() {
+        EntityWithEmbeddedIdWithFinalField persistedEntity = new EntityWithEmbeddedIdWithFinalField();
+        persistedEntity.setName("Some name");
+        inTransaction(() -> {
+            em.persist(persistedEntity);
+        });
+
+        // Read with the same ID instance
+        inTransaction(() -> {
+            EntityWithEmbeddedIdWithFinalField entity = em.find(EntityWithEmbeddedIdWithFinalField.class,
+                    persistedEntity.getId());
+            assertThat(entity).extracting(EntityWithEmbeddedIdWithFinalField::getId).extracting(i -> i.id)
+                    .isEqualTo(persistedEntity.getId().id);
+        });
+
+        // Read with a new ID instance
+        inTransaction(() -> {
+            EntityWithEmbeddedIdWithFinalField entity = em.find(EntityWithEmbeddedIdWithFinalField.class,
+                    EntityWithEmbeddedIdWithFinalField.EmbeddableId.of(persistedEntity.getId().id));
+            assertThat(entity).extracting(EntityWithEmbeddedIdWithFinalField::getId).extracting(i -> i.id)
+                    .isEqualTo(persistedEntity.getId().id);
+        });
+
+        // Read with a query
+        // This is special because in this particular test,
+        // we know Hibernate ORM *has to* instantiate the EmbeddableIdType itself:
+        // it cannot reuse the ID we passed.
+        // And since the EmbeddableIdType has a final field, instantiation will not be able to use a no-arg constructor...
+        inTransaction(() -> {
+            EntityWithEmbeddedIdWithFinalField entity = em
+                    .createQuery("from embidwithfinal e where e.name = :name", EntityWithEmbeddedIdWithFinalField.class)
+                    .setParameter("name", persistedEntity.getName())
+                    .getSingleResult();
+            assertThat(entity).extracting(EntityWithEmbeddedIdWithFinalField::getId).extracting(i -> i.id)
+                    .isEqualTo(persistedEntity.getId().id);
+        });
+    }
+
+    @Test
+    public void embeddableNonIdWithFinalField_constructor() {
+        EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId embeddable = new EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId(
+                "foo");
+        assertThat(embeddable.immutableProperty).isEqualTo("foo");
+    }
+
+    // Just test that the embedded non-ID works correctly over a persist/retrieve cycle
+    @Test
+    public void embeddableNonIdWithFinalField_smokeTest() {
+        EntityWithEmbeddedNonIdWithFinalField persistedEntity = new EntityWithEmbeddedNonIdWithFinalField();
+        persistedEntity.setName("Some name");
+        persistedEntity.setEmbedded(new EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId("foo"));
+        inTransaction(() -> {
+            em.persist(persistedEntity);
+        });
+
+        inTransaction(() -> {
+            EntityWithEmbeddedNonIdWithFinalField entity = em.find(EntityWithEmbeddedNonIdWithFinalField.class,
+                    persistedEntity.getId());
+            assertThat(entity).extracting(EntityWithEmbeddedNonIdWithFinalField::getEmbedded)
+                    .extracting(EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId::getImmutableProperty)
+                    .isEqualTo(persistedEntity.getEmbedded().getImmutableProperty());
+        });
+    }
+
+    private void inTransaction(Runnable runnable) {
+        try {
+            transaction.begin();
+            try {
+                runnable.run();
+                transaction.commit();
+            } catch (Exception e) {
+                transaction.rollback();
+            }
+        } catch (SystemException | NotSupportedException e) {
+            throw new IllegalStateException("Transaction exception", e);
+        }
+    }
+
+    @Entity(name = "withfinal")
+    public static class EntityWithFinalField {
+
+        @Id
+        @GeneratedValue
+        private Long id;
+
+        private final String immutableProperty;
+
+        private String name;
+
+        // For Hibernate ORM
+        protected EntityWithFinalField() {
+            this.immutableProperty = null;
+        }
+
+        private EntityWithFinalField(String id) {
+            this.immutableProperty = id;
+        }
+
+        public Long getId() {
+            return id;
+        }
+
+        public void setId(Long id) {
+            this.id = id;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public String getImmutableProperty() {
+            return immutableProperty;
+        }
+    }
+
+    @Entity(name = "embidwithfinal")
+    public static class EntityWithEmbeddedIdWithFinalField {
+
+        @EmbeddedId
+        private EmbeddableId id;
+
+        private String name;
+
+        public EntityWithEmbeddedIdWithFinalField() {
+            this.id = EmbeddableId.of(UUID.randomUUID().toString());
+        }
+
+        public EmbeddableId getId() {
+            return id;
+        }
+
+        public void setId(EmbeddableId id) {
+            this.id = id;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        @Immutable
+        @Embeddable
+        public static class EmbeddableId implements Serializable {
+            private final String id;
+
+            // For Hibernate ORM
+            protected EmbeddableId() {
+                this.id = null;
+            }
+
+            private EmbeddableId(String id) {
+                this.id = id;
+            }
+
+            public static EmbeddableId of(String string) {
+                return new EmbeddableId(string);
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) {
+                    return true;
+                }
+                if (!(o instanceof EmbeddableId)) {
+                    return false;
+                }
+                EmbeddableId embeddableIdType = (EmbeddableId) o;
+                return Objects.equals(id, embeddableIdType.id);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(id);
+            }
+        }
+    }
+
+    @Entity(name = "embwithfinal")
+    public static class EntityWithEmbeddedNonIdWithFinalField {
+
+        @Id
+        @GeneratedValue
+        private Long id;
+
+        private String name;
+
+        @Embedded
+        private EmbeddableNonId embedded;
+
+        public Long getId() {
+            return id;
+        }
+
+        public void setId(Long id) {
+            this.id = id;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public EmbeddableNonId getEmbedded() {
+            return embedded;
+        }
+
+        public void setEmbedded(
+                EmbeddableNonId embedded) {
+            this.embedded = embedded;
+        }
+
+        @Embeddable
+        public static class EmbeddableNonId {
+            private final String immutableProperty;
+
+            private String mutableProperty;
+
+            protected EmbeddableNonId() {
+                // For Hibernate ORM only - it will change the property value through reflection
+                this.immutableProperty = null;
+            }
+
+            private EmbeddableNonId(String immutableProperty) {
+                this.immutableProperty = immutableProperty;
+            }
+
+            public String getImmutableProperty() {
+                return immutableProperty;
+            }
+
+            public String getMutableProperty() {
+                return mutableProperty;
+            }
+
+            public void setMutableProperty(String mutableProperty) {
+                this.mutableProperty = mutableProperty;
+            }
+        }
+    }
+}

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/publicfields/PublicFieldAccessFinalFieldTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/publicfields/PublicFieldAccessFinalFieldTest.java
@@ -1,0 +1,260 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package io.quarkus.hibernate.orm.publicfields;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.UUID;
+
+import javax.inject.Inject;
+import javax.persistence.Embeddable;
+import javax.persistence.Embedded;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import javax.persistence.EntityManager;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.transaction.NotSupportedException;
+import javax.transaction.SystemException;
+import javax.transaction.UserTransaction;
+
+import org.hibernate.annotations.Immutable;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Checks that public, final field access is correctly replaced with getter calls for reads,
+ * but not replaced at all for writes (since writes to final fields can only occur from constructors).
+ * <p>
+ * See https://github.com/quarkusio/quarkus/issues/20186
+ */
+public class PublicFieldAccessFinalFieldTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(
+                            EntityWithFinalField.class,
+                            EntityWithEmbeddedIdWithFinalField.class, EntityWithEmbeddedIdWithFinalField.EmbeddableId.class,
+                            EntityWithEmbeddedNonIdWithFinalField.class,
+                            EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId.class))
+            .withConfigurationResource("application.properties");
+
+    @Inject
+    EntityManager em;
+
+    @Inject
+    UserTransaction transaction;
+
+    @Test
+    public void entityWithFinalField_constructor() {
+        EntityWithFinalField entity = new EntityWithFinalField("foo");
+        assertThat(entity.immutableProperty).isEqualTo("foo");
+    }
+
+    // Just test that the embedded non-ID works correctly over a persist/retrieve cycle
+    @Test
+    public void entityWithFinalField_smokeTest() {
+        EntityWithFinalField persistedEntity = new EntityWithFinalField("foo");
+        persistedEntity.name = "Some name";
+        inTransaction(() -> {
+            em.persist(persistedEntity);
+        });
+
+        inTransaction(() -> {
+            EntityWithFinalField entity = em.find(EntityWithFinalField.class, persistedEntity.id);
+            assertThat(entity).extracting(e -> e.immutableProperty)
+                    .isEqualTo(persistedEntity.immutableProperty);
+        });
+    }
+
+    // Just test that the embedded ID works correctly over a persist/retrieve cycle
+    @Test
+    public void embeddableIdWithFinalField_smokeTest() {
+        EntityWithEmbeddedIdWithFinalField persistedEntity = new EntityWithEmbeddedIdWithFinalField();
+        persistedEntity.name = "Some name";
+        inTransaction(() -> {
+            em.persist(persistedEntity);
+        });
+
+        // Read with the same ID instance
+        inTransaction(() -> {
+            EntityWithEmbeddedIdWithFinalField entity = em.find(EntityWithEmbeddedIdWithFinalField.class,
+                    persistedEntity.id);
+            assertThat(entity).extracting(e -> e.id).extracting(i -> i.id)
+                    .isEqualTo(persistedEntity.id.id);
+        });
+
+        // Read with a new ID instance
+        inTransaction(() -> {
+            EntityWithEmbeddedIdWithFinalField entity = em.find(EntityWithEmbeddedIdWithFinalField.class,
+                    EntityWithEmbeddedIdWithFinalField.EmbeddableId.of(persistedEntity.id.id));
+            assertThat(entity).extracting(e -> e.id).extracting(i -> i.id)
+                    .isEqualTo(persistedEntity.id.id);
+        });
+
+        // Read with a query
+        // This is special because in this particular test,
+        // we know Hibernate ORM *has to* instantiate the EmbeddableIdType itself:
+        // it cannot reuse the ID we passed.
+        // And since the EmbeddableIdType has a final field, instantiation will not be able to use a no-arg constructor...
+        inTransaction(() -> {
+            EntityWithEmbeddedIdWithFinalField entity = em
+                    .createQuery("from embidwithfinal e where e.name = :name", EntityWithEmbeddedIdWithFinalField.class)
+                    .setParameter("name", persistedEntity.name)
+                    .getSingleResult();
+            assertThat(entity).extracting(e -> e.id).extracting(i -> i.id)
+                    .isEqualTo(persistedEntity.id.id);
+        });
+    }
+
+    @Test
+    public void embeddableNonIdWithFinalField_constructor() {
+        EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId embeddable = new EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId(
+                "foo");
+        assertThat(embeddable.immutableProperty).isEqualTo("foo");
+    }
+
+    // Just test that the embedded non-ID works correctly over a persist/retrieve cycle
+    @Test
+    public void embeddableNonIdWithFinalField_smokeTest() {
+        EntityWithEmbeddedNonIdWithFinalField persistedEntity = new EntityWithEmbeddedNonIdWithFinalField();
+        persistedEntity.name = "Some name";
+        persistedEntity.embedded = new EntityWithEmbeddedNonIdWithFinalField.EmbeddableNonId("foo");
+        inTransaction(() -> {
+            em.persist(persistedEntity);
+        });
+
+        inTransaction(() -> {
+            EntityWithEmbeddedNonIdWithFinalField entity = em.find(EntityWithEmbeddedNonIdWithFinalField.class,
+                    persistedEntity.id);
+            assertThat(entity).extracting(e -> e.embedded)
+                    .extracting(emb -> emb.immutableProperty)
+                    .isEqualTo(persistedEntity.embedded.immutableProperty);
+        });
+    }
+
+    private void inTransaction(Runnable runnable) {
+        try {
+            transaction.begin();
+            try {
+                runnable.run();
+                transaction.commit();
+            } catch (Exception e) {
+                transaction.rollback();
+            }
+        } catch (SystemException | NotSupportedException e) {
+            throw new IllegalStateException("Transaction exception", e);
+        }
+    }
+
+    @Entity(name = "withfinal")
+    public static class EntityWithFinalField {
+
+        @Id
+        @GeneratedValue
+        public Long id;
+
+        public final String immutableProperty;
+
+        public String name;
+
+        // For Hibernate ORM
+        protected EntityWithFinalField() {
+            this.immutableProperty = null;
+        }
+
+        private EntityWithFinalField(String id) {
+            this.immutableProperty = id;
+        }
+
+    }
+
+    @Entity(name = "embidwithfinal")
+    public static class EntityWithEmbeddedIdWithFinalField {
+
+        @EmbeddedId
+        public EmbeddableId id;
+
+        public String name;
+
+        public EntityWithEmbeddedIdWithFinalField() {
+            this.id = EmbeddableId.of(UUID.randomUUID().toString());
+        }
+
+        @Immutable
+        @Embeddable
+        public static class EmbeddableId implements Serializable {
+            public final String id;
+
+            // For Hibernate ORM
+            protected EmbeddableId() {
+                this.id = null;
+            }
+
+            private EmbeddableId(String id) {
+                this.id = id;
+            }
+
+            public static EmbeddableId of(String string) {
+                return new EmbeddableId(string);
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) {
+                    return true;
+                }
+                if (!(o instanceof EmbeddableId)) {
+                    return false;
+                }
+                EmbeddableId embeddableIdType = (EmbeddableId) o;
+                return Objects.equals(id, embeddableIdType.id);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(id);
+            }
+        }
+    }
+
+    @Entity(name = "embwithfinal")
+    public static class EntityWithEmbeddedNonIdWithFinalField {
+
+        @Id
+        @GeneratedValue
+        public Long id;
+
+        public String name;
+
+        @Embedded
+        public EmbeddableNonId embedded;
+
+        @Embeddable
+        public static class EmbeddableNonId {
+            public final String immutableProperty;
+
+            public String mutableProperty;
+
+            protected EmbeddableNonId() {
+                // For Hibernate ORM only - it will change the property value through reflection
+                this.immutableProperty = null;
+            }
+
+            private EmbeddableNonId(String immutableProperty) {
+                this.immutableProperty = immutableProperty;
+            }
+        }
+    }
+}

